hathor/indexes/base_index.py

```diff
@@ -13,12 +13,82 @@
+    topological_order: bool = True  # if False than ordering doesn't matter
+
+    # XXX: this is used to join the scope of multiple indexes to get an overall scope that includes everything that
+    #      each individual scope needs, the OR operator was chosen because it represents well the operation of keeping
+    #      a property if either A or B needs it
+    def __or__(self, other):
+        # XXX: note that this doesn't necessarily have to be OR operations between properties, we want the operations
+        #      that broaden the scope, and not narrow it.
+        # XXX: in the case of topological_order, we want to keep the "topological" ordering if any of them requires it,
+        #      so it also is an OR operator
+        return Scope(
+            include_blocks=self.include_blocks | other.include_blocks,
+            include_txs=self.include_txs | other.include_txs,
+            include_voided=self.include_voided | other.include_voided,
+            include_partial=self.include_partial | other.include_partial,
+            topological_order=self.topological_order | other.topological_order,
+        )
+
+    def matches(self, tx: BaseTransaction) -> bool:
+        """ Check if a transaction matches this scope, True means the index is interested in this transaction.
+        """
+        if tx.is_block and not self.include_blocks:
+            return False
+        if tx.is_transaction and not self.include_txs:
+            return False
+        tx_meta = tx.get_metadata()
+        if tx_meta.voided_by and not self.include_voided:
+            return False
+        if not tx_meta.validation.is_fully_connected() and not self.include_partial:
+            return False
+        # XXX: self.topologial_order doesn't affect self.match()
+        # passed all checks
+        return True
+
+    def get_iterator(self, tx_storage: 'TransactionStorage') -> Iterator[BaseTransaction]:
+        iterator: Iterator[BaseTransaction]
+        # XXX: this is to mark if the chosen iterator will yield partial transactions
+        iterator_covers_partial: bool
+        if self.topological_order:
+            iterator = tx_storage.topological_iterator()
+            iterator_covers_partial = False
+        else:
+            iterator = tx_storage.get_all_transactions()
+            iterator_covers_partial = True
+        for tx in iterator:
+            if self.matches(tx):
+                yield tx
+        if self.include_partial and not iterator_covers_partial:
+            # if partial transactions are needed and were not already covered, we use get_all_transactions, which
+            # includes partial transactions, to yield them, skipping all that aren't partial
+            for tx in tx_storage.get_all_transactions():
+                tx_meta = tx.get_metadata()
+                if tx_meta.validation.is_fully_connected():
+                    continue
+                if self.matches(tx):
+                    yield tx
```

Comment:
If `self.topological_order = True` implies requiring topological order, is it correct to yield all partial transactions after the ones from the `topological_iterator`? In other words, are partial transactions always considered last in the topological order?

Reply:
Exactly, partial transactions will always come after all the other transactions that were yielded in topological order first, mainly because it's the only straightforward way to implement it (otherwise we'd need to index them, and having indexes depend on each other adds complexity that we don't need to add right now, besides the deps-index already handles knowing all partially validated transactions). There's no other option, since there can't be a partially validated transaction between two fully validated transactions (or rather, there _could_ be, but there must always be a topological order where there isn't one, since fully validated transactions will always point to other fully validated transactions). Also partially validated transactions don't _really_ need to be sorted, specially because it's not guaranteed that we'll be able to sort them.

Reply:
Got it. Thanks!